pkg/hostpath/nodeserver.go

```diff
@@ -65,22 +72,51 @@ func getPodDetails(volumeContext map[string]string) (string, string, string, str
 
 }
 
-func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolumeRequest) (*csi.NodePublishVolumeResponse, error) {
-	var targetPath string
+func (ns *nodeServer) validateShare(req *csi.NodePublishVolumeRequest) (*sharev1alpha1.Share, error) {
+	shareName, sok := req.GetVolumeContext()[ProjectedResourceShareKey]
+	if !sok || len(strings.TrimSpace(shareName)) == 0 {
+		return nil, status.Errorf(codes.InvalidArgument,
+			"the csi driver reference is missing the volumeAttribute 'share'")
+	}
 
-	// Check arguments
-	if req.GetVolumeCapability() == nil {
-		return nil, status.Error(codes.InvalidArgument, "Volume capability missing in request")
+	share, err := client.GetListers().Shares.Get(shareName)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument,
+			"the csi driver volumeAttribute 'share' reference had an error: %s", err.Error())
 	}
-	if len(req.GetVolumeId()) == 0 {
-		return nil, status.Error(codes.InvalidArgument, "Volume ID missing in request")
+
+	switch strings.TrimSpace(share.Spec.BackingResource.Kind) {
+	case "Secret":
+	case "ConfigMap":
+	default:
+		return nil, status.Errorf(codes.InvalidArgument,
+			"the share %s has an invalid backing resource kind %s", shareName, share.Spec.BackingResource.Kind)
 	}
-	if len(req.GetTargetPath()) == 0 {
-		return nil, status.Error(codes.InvalidArgument, "Target path missing in request")
+
+	if len(strings.TrimSpace(share.Spec.BackingResource.Namespace)) == 0 {
+		return nil, status.Errorf(codes.InvalidArgument,
+			"the share %s backing resource namespace needs to be set", shareName)
 	}
-	if req.VolumeContext == nil || len(req.GetVolumeContext()) == 0 {
-		return nil, status.Error(codes.InvalidArgument, "Volume attributes missing in request")
+	if len(strings.TrimSpace(share.Spec.BackingResource.Name)) == 0 {
+		return nil, status.Errorf(codes.InvalidArgument,
+			"the share %s backing resource name needs to be set", shareName)
+	}
+
+	podNamespace, podName, _, podSA := getPodDetails(req.GetVolumeContext())
+
+	allowed, err := client.ExecuteSAR(shareName, podNamespace, podName, podSA)
+	if allowed {
+		return share, nil
 	}
+	return nil, err
+}
+
+// validateVolumeContext return values:
+// - podNamespace
+// - podNames
+// - podUID
+// - podSA
+func (ns *nodeServer) validateVolumeContext(req *csi.NodePublishVolumeRequest) (string, string, string, string, error) {
```

Comment:
Why do we need to return the pod data when this information exists in the `NodePublishVolumeRequest`?

Returning four `strings` plus an `error` feels like a golang code smell.

Reply:
what, multiple returns is one of the joys of golang :-)

I'll make an attempt at restructuring to this to facilitate the various concerns, including what you've noted, and see if it sticks at least for me.

Reply:
OK it accesses the vol ctx an time or two more than before, but the organization nodeserver vs. hostpath is much better, and among other things this methods returns just an error now @adambkaplan 

thanks